### PR TITLE
GC Live Object Check

### PR DESCRIPTION
**Pull request by fh-ms**
Add live object check to GC. Ensures retaining of objects in storage when still live in VM. Additional regression concurrency fix.
Finally fixes all remaining GC issues, brings it out of the experimental stage and activates it by default.

### CODE REVIEW

File: base/src/main/java/one/microstream/collections/Set_long.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
+package one.microstream.collections;
+
+import one.microstream.collections.interfaces.OptimizableCollection;
+import one.microstream.functional._longIterable;
+import one.microstream.functional._longPredicate;
+import one.microstream.functional._longProcedure;
+import one.microstream.math.XMath;
+import one.microstream.typing.Composition;
+
+/**
```
Comment (zdenek-jonas) [inline base/src/main/java/one/microstream/collections/Set_long.java]: empty javadoc?

Comment (fh-ms) [reply base/src/main/java/one/microstream/collections/Set_long.java]: Fixed

File: persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java [hunk 2]
```diff
@@ -162,6 +172,20 @@ public static DefaultObjectRegistry New(
 	// instance fields //
 	////////////////////
 	
+	/*
+	 * Note:
+	 * This does NOT replace locking the whole registry over a not to be "disrupted" process like loading
+	 * (See BinaryLoader#get).
+	 * This is just an internal lock to keep things consistent on a technical level.
+	 * Locking the registry itself continously accross a whole process keeps things consistent
```
Comment (hg-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java]: typo 'continously accross'

File: persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java [hunk 3]
```diff
@@ -972,18 +1084,48 @@ private void internalBuildConstantsColdStorage()
...
+		synchronized(this.mutex)
+		{
+			processor.processObjectIdsByFilter(this::synchIsLiveObjectId);
+			return true;
+		}
+	}
+
+	final boolean synchIsLiveObjectId(final long objectId)
+	{
+		final boolean result = this.synchContainsObjectId(objectId);
```
Comment (hg-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java:1103]: maybe remove commented code or add logging

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		@Override
+		public synchronized void registerStorer(final PersistenceStorer storer)
+		{
+			HashEnum<PersistenceStorer> storerGroup = this.storerGroups.get(this.currentGroupId);
+			if(storerGroup == null)
+			{
+				this.storerGroups.add(this.currentGroupId, storerGroup = HashEnum.New());
+			}
+
+//			XDebug.print("Registering storer " + XChars.systemString(storer) + " to id Group " + this.currentGroupId);
```
Comment (hg-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java]: maybe add logging or remove?

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+			storerGroup.add(storer);
+		}
+
+		@Override
+		public synchronized boolean clearGroupAndAdvance(final long oldGroupId, final long newGroupId)
+		{
+			final long removeCount = this.storerGroups.removeBy(e -> e.key() <= oldGroupId);
+
+//			XDebug.println(Thread.currentThread() + " removed " + removeCount + " idGroups with id <= " + oldGroupId + ".");
```
Comment (hg-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java]: maybe add logging or remove?

**Review by zdenek-jonas — APPROVED** [review]

File: storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageConnectionFoundation.java [hunk 6]
```diff
@@ -221,6 +302,40 @@ protected StorageRequestAcceptor internalGetStorageRequestAcceptor()
...
+		}
+
+		@Override
+		public PersistenceManager<Binary> createPersistenceManager()
+		{
+			final PersistenceLiveStorerRegistry storerRegistry = this.getLiveStorerRegistry();
+			final PersistenceStorer.CreationObserver observer = this.getStorerCreationObserver();
+			if(observer == null)
+			{
+				// registry can simply be set as the (sole) obserber
```
Comment (hg-ms) [inline storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageConnectionFoundation.java]: typo in 'obserber'

File: storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageFoundation.java [hunk 7]
```diff
@@ -728,6 +730,13 @@ protected StorageObjectIdRangeEvaluator ensureObjectIdRangeEvaluator()
 				}
 			};
 		}
+		
+		@Override
+		protected EmbeddedStorageObjectRegistryCallback ensureObjectIdsSelector()
+		{
+			return this.getConnectionFoundation().getObjectRegistryCallback();
+//			return EmbeddedStorageObjectRegistryCallback.New();
```
Comment (hg-ms) [inline storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageFoundation.java]: remove commented code

File: storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageObjectRegistryCallback.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		public synchronized void initializeObjectRegistry(final PersistenceObjectRegistry objectRegistry)
+		{
+			if(this.objectRegistry != null)
+			{
+				if(this.objectRegistry == objectRegistry)
+				{
+					return;
+				}
+
+				// (29.07.2022 TM)EXCP: proper exception
```
Comment (hg-ms) [inline storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageObjectRegistryCallback.java:73]: is this a TODO?

Comment (fh-ms) [reply storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageObjectRegistryCallback.java:73]: All places where RuntimeExceptions should be replaced by properly typed ones are marked with this expression. It is about time to replace them; see #403.

File: persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java (hunk 2 above)
Comment (fh-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java]: Fixed

File: persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java (hunk 3 above)
Comment (fh-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/internal/DefaultObjectRegistry.java:1103]: Added logging

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java (hunk 4 above)
Comment (fh-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java]: Added logging

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java (hunk 5 above)
Comment (fh-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceLiveStorerRegistry.java]: Added logging

File: storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageConnectionFoundation.java (hunk 6 above)
Comment (fh-ms) [reply storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageConnectionFoundation.java]: Fixed

File: storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageFoundation.java (hunk 7 above)
Comment (fh-ms) [reply storage/embedded/src/main/java/one/microstream/storage/embedded/types/EmbeddedStorageFoundation.java]: Removed

**Review by hg-ms — APPROVED** [review]

**Pull request merged**